### PR TITLE
feat(Core/CreatureAI): improve npc position during the combat

### PR DESCRIPTION
**Pull request by Helias**
Changed the creature's positions while attacking the player.

Authors: 
@Helias 
@Yehonal
@lucaAngrisani

PR inspired by:
@Langerz82 ( https://gist.github.com/callmephil/5c5d18e1db0ca4e86115a299b836cabf )

## ISSUES ADDRESSED:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/3357

## EXPECTED BLIZZLIKE BEHAVIOR
The first four videos of this playlist
https://www.youtube.com/playlist?list=PLx1yhYXFOxyFCrLWJIr2dfyySx5jiEr4k

## HOW THIS PR WORK
last video of the playlist
https://www.youtube.com/watch?v=ogmPCycMC3U&list=PLx1yhYXFOxyFCrLWJIr2dfyySx5jiEr4k&index=5

## HOW TO TEST
- Attack a group of NPCs and see how they reposition to attack you.
- Later, go really close to the NPCs they should go back a bit to attack you

Be careful:
- inside the dungeons, there is more "tolerance" about reposition, so they can also not reposition instantly
- outside the dungeons, it's easy that the NPCs reposition themselves.

## Target branch(es):
- [x] Master

## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by felwyn** [comment]
Fix is working. NPC will now re-position like the picture below
WoWScrnShot_082920_153552
WoWScrnShot_082920_153345

**Comment by Helias** [comment]
Thank you @felwyn!

I want to be sure about this script, so I would like that other testers test this PR.

**Comment by LukasVolgger** [comment]
Tested. I agree with @felwyn

**Comment by locus313** [comment]
tested, looks good.

**Comment by locus313** [comment]
@felwyn @LukasVolgger did you do any testing in any instances? i was doing some further testing and saw some mobs now glitch out and floating when they should not.

**Comment by felwyn** [comment]
Okay I am back soon and let me check this furthermore. Thanks for the info.

**Comment by FrancescoBorzi** [comment]
@locus313

 > i was doing some further testing and saw some mobs now glitch out and floating when they should not.

is this a regression?  (regression = they work better on master, but this PR makes it worse)

**Comment by felwyn** [comment]
no this Fix was doing its purpose. Mobs are circling around and/or repositioning when you stand in the middle of their "point". I guess the code needs still a bit of optimization.

I will be a bit delayed with testing because of RL but I will confirm/not confirm this issue later this day

**Comment by locus313** [comment]
@FrancescoBorzi Yes this does look like a regression. Here is a screen shot from testing this pr in the nexus:
image
I did not see this behavior on the master branch.

**Comment by locus313** [comment]
Here is another couple screenshots from mana-tombs where the mob is falling through the floor:
image
image

**Comment by felwyn** [comment]
Wait a second. 

I had this bug appearing when I tried to kill Prince Sandoval (the bug unattackable). I died accidentally and I had that Spririt Healer Geivvhr having the exact same issue. He was clipping trought the ground until I abandoned the quest and retried it.

**Comment by Helias** [comment]
I will try to add this AI feature only outside the building (and outside dungeon/instance), to apply it only in a open space zone and restrict it

**Comment by felwyn** [comment]
but when we fight indoors and in instances we will have the original issue. That is not really the best way to fix the problem.

**Comment by FrancescoBorzi** [comment]
I agree with @felwyn , we should find a proper solution for this

**Comment by Helias** [comment]
I am not sure, but probably I fixed it.
We need several tests, really, deep test.

Especially, I think that we should test this kind of movement climbing a mountains or some stairs.
image

**Comment by Helias** [comment]
pushed new edits, let's test!

**Comment by locus313** [comment]
tested, ran through a couple instances. no longer see the issue i saw before.

**Comment by FrancescoBorzi** [comment]
thanks @locus313 , however let's test this a bit more just to be safe. We don't want to break stuff :) Can you keep playing with this for a while?

also @azerothcore/testers  can you also give this a try?

**Comment by locus313** [comment]
@FrancescoBorzi yep, played several more instances today. ill do some more testing on the pr over the next several days.

**Comment by Helias** [comment]
Try also to flight and take aggro of some NPCs while you are flying.
Try to fight climbing a mountain

**Comment by Helias** [comment]
During the fight, the NPCs and the bosses take distance while they cast a spell, is it normal? I think that this PR affect this behavior

**Comment by locus313** [comment]
@Helias Your right, This behavior changed with this pr.  I don't think the npc or bosses should take distance before casting a spell. I did also check some youtube videos and didn't see behavior like that. So does look like this pr has affect that.

**Comment by locus313** [comment]
@Helias tested climbing a mountain. mobs do end up stacking sometimes:
image

**Comment by locus313** [comment]
while flying the mobs stack for about a second then spread out slightly:
image

**Comment by craftorix** [comment]
Does this fix have to apply for NPC fights too? Because it doesnt apply when only NPCs are fighting.

**Comment by Helias** [comment]
This PR should not affect the fighting between NPCs

**Comment by markvetwow** [comment]
@locus313 Since you've mentioned this, i can say that i can detect this problem too on my server. It was happening and happening still even before this PR. Mostly in dungeons. Let's take Kara for example. Ghostly Guests will take a great distance before they start casting something, even if they were interrupted they can stay still and wait for the next cast. If you're gonna make a closer distance, they start to ran away from you. This can happen to Moroes guests as well.

**Comment by Helias** [comment]
Updated. I restored the old way to take distance when casting, so this PR will not affect the old movement about taking distance, but ofc, it will keep the way to distribute the NPCs around the character

**Comment by locus313** [comment]
@Helias the casters will start casting without moving back, but then after a little bit the caster will move back then come forward again. Here is a couple screenshoots:
image
image

**Comment by markvetwow** [comment]
Since there is a caster's movement discussion, wanted to share some strange behaviour from some casters (not only casters to be clear, this behaviour can be seen even on the ranged mobs, like shooters in Old Hillsbrad for example).[ Link to the video](https://www.youtube.com/watch?v=qpgMHPZDyBU).

When you get in melee range, NPC is trying to move away from you to a needed distance before they start doing smth.

**Comment by Helias** [comment]
I removed any reference about back-movement, now it should work as we expect :P

@locus313, thank you for testing until now, could you make this (hope last) test? :P

**Comment by locus313** [comment]
@Helias yep, ill test it again shortly

**Comment by locus313** [comment]
@Helias When you first walk up to the casters they now do not automatically back up before casting. But found something interesting. If i walk backwards from the caster a little bit they will then move back before starting to cast again. Then they will come closer again. Screenshot from this test:
image

**Comment by pklloveyou** [comment]
It doesn't need to be so much trouble，It can be realized, as shown in the figure
image

**Comment by pklloveyou** [comment]
image
image
Stop when I run away

**Comment by Helias** [comment]
I don't think that this PR affect this, could you check if it happens on master too?

**Comment by locus313** [comment]
@Helias was not able to reproduce this on the master branch.

**Comment by markvetwow** [comment]
Tested this today again, there is a problem that mobs are teleporting to current aggro target when they are rooted. (Frost Nova for example).

**Comment by markvetwow** [comment]
No news on this issue?

**Comment by Helias** [comment]
Sorry but I am a bit busy in this period :(

**Comment by Helias** [comment]
I cleaned up the code and excluded this feature to all Scripted NPCs, this feature will be applied only to NPCs that have no `AIName` and `ScriptName` assigned.
This will exclude all the bosses and also all the SmartAI and any kind of *AI NPC.
It's a bit a pity for several NPCs that should be involved in this feature but it's also "safer" in this way to do not break any specific scripted NPC.

Therefore, only simple NPCs like these ones will be affected by this feature.
image


But not these ones:
image

Let's test it!

**Comment by Yehonal** [comment]
Could someone test NPCs that have the 4 spells defined in
creature_templare? How they act.
But also NPC with scriptAI but of melee type

Il ven 25 dic 2020, 13:24 Stefano Borzì <notifications@github.com> ha
scritto:

**Comment by Helias** [comment]
Tested with 1839, it seems working :P
This NPC has 4 spell, it is a `rare` NPC
image

**Comment by Helias** [comment]
I've just found a little bug, if I fall from a mountain the NPCs can also fall THROUGH the texture and come back to their original position, I can't reproduce this on master, I should investigate to fix it >.>.
I think it's a minor issue, but may if someone test it personally can give me a better feedback.

**Comment by Yehonal** [comment]
Do they cast the 4 spells?

Il ven 25 dic 2020, 16:22 Stefano Borzì <notifications@github.com> ha
scritto:

**Comment by Yehonal** [comment]
I'm not sure if it can be helpful but this is what happens on Classic (offy) right now: https://youtu.be/jzYGE9Dls8c

I would like to compare also with NPCs that cast spells asap

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -2189,6 +2226,98 @@ void Unit::AttackerStateUpdate (Unit* victim, WeaponAttackType attType, bool ext
...
+            refUnit = otherAttacher;
+        }
+    }
+
+    if (!refUnit || meleeReach <= 0)
+        return NULL;
+
+    double refUnitRay = refUnit->GetObjectSize() / 2.0f;
+
+    // Equation of tangent point
```
Comment (Yehonal) [inline src/server/game/Entities/Unit/Unit.cpp:2273]: Thanks to @Lucangrisani for this

**Comment by Yehonal** [comment]
I've pushed the last version of the algorithm, it resembles perfectly the retail behaviours, including the troubles of the mechanics (such as the endless positioning when there's no space in outworld)

All my tests and research have been published here: https://www.youtube.com/playlist?list=PLx1yhYXFOxyFCrLWJIr2dfyySx5jiEr4k

Together with @lucaangrisani we studied a formula to allow the creatures moving outside of collisions with a very precise angle

image

Here you can see the idea behind.

You can see the formula here: https://github.com/azerothcore/azerothcore-wotlk/pull/3369/files#diff-b4a23722c5b449d326b1c0e5a75f3b51de931e42e688793fd9f1d233536a02c6R2285


This PR also includes the "backward" functionality using a proper angle and a dynamic collision tolerance in instance

File: data/sql/updates/db_world/2020_12_25_00.sql [hunk 2]
```diff
@@ -0,0 +1,629 @@
+-- DB update 2020_12_24_01 -> 2020_12_25_00
```
Comment (FrancescoBorzi) [inline data/sql/updates/db_world/2020_12_25_00.sql:1]: should use pending SQL to add new SQL code

File: src/server/game/Entities/Unit/Unit.cpp [hunk 3]
```diff
@@ -2242,124 +2232,90 @@ struct AttackDistance {
...
-    Position const& pos = GetPosition();
+    double refUnitRay = refUnit->GetObjectSize() / 2.0f;
 
-    uint8 i = 0;
-    for (const auto& attacker: meleeAttackers)
-    {
-        // CreatureModelDataEntry const* modelData = sCreatureModelDataStore.LookupEntry(attacker->GetDisplayId());
-        // float size = modelData->Scale ? modelData->Scale : 1;
+    // Equation of tangent point
+    if (GetExactDist2d(refUnit) > refUnitRay) { // works only when there's enough space
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Unit/Unit.cpp]: codestyle

File: src/server/game/Entities/Unit/Unit.cpp [hunk 4]
```diff
@@ -2242,124 +2232,90 @@ struct AttackDistance {
...
-            meleeAttackers.insert(attacker);
+        if (otherAttacher->GetGUID() == attacker->GetGUID() || otherAttacher->isMoving())
+            continue;
+
+        float tempDist = attacker->GetExactDist2d(otherAttacher) - (attacker->GetObjectSize()/2) - (otherAttacher->GetObjectSize()/2);
+
+        if (tempDist > distanceTollerance)
+            continue;
+
+        if (tempDist == 0 || minDistance == 0 || tempDist < minDistance) {
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Unit/Unit.cpp:2265]: codestyle

File: src/server/game/Entities/Unit/Unit.h [hunk 5]
```diff
@@ -2451,6 +2473,8 @@ class Unit : public WorldObject
     // Movement info
     Movement::MoveSpline* movespline;
 
+    AttackPosition getAttackPosition() { return m_attackPosition; }
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Unit/Unit.h:2476]: use `[[nodiscard]]` where appropriate

File: data/sql/updates/db_world/2020_12_25_00.sql (hunk 2 above)
Comment (Yehonal) [reply data/sql/updates/db_world/2020_12_25_00.sql:1]: we don't have to add SQL code here...idk why there are such files there

**Comment by Helias** [comment]
The PR has been reworked, now it's ready to TEST.
We did a big job so I ask you all to test this pls.
This PR will affect ALL NPCs in the world, this is a feature that from level 1 is present in the game, in the PR I linked the **video made by Yehonal that compare the retail behavior of the NPCs and the current behavior with this PR.**

I also made several tests like:
- Mana-tomb NPCs
- Onyxia boss where there are multiple little dragons spawned from her and also the boss needs to fly away for a big period and also later I tested it with their little dragons next-to.
- several NPCs BIG and SMALL, because this PR calculate the repositioning considering also the SIZE of the NPCs

The PR has been tested successfully but I want that at least other 2 testers try to test **one entire instance** and give some feedback.

@locus313 @LukasVolgger @pangolp @nicoaravena
@azerothcore/testers

**Comment by Yehonal** [comment]
Just for the record, this is the issue on Dalaran WoW: https://github.com/dalaranwow/dalaran-wow/issues/4711 You can get more info here.

They are probably using TC, but the issue was the same

**Comment by Helias** [comment]
My tests:

<details>
  <summary>Show tests</summary>
 
image
image
image

I did more tests but I didn't make the screen for everything, just trust me xD

</details>

**Comment by Langerz82** [comment]
Expect these wankers will tear apart ur code they are certified fags, happy NY.

**Comment by Yehonal** [comment]
> Expect these wankers will tear apart ur code they are certified fags, happy NY.

Which code?

**Comment by nicoaravena** [comment]
Testing results:
Doing the Nexus
imagen

Noticed that Mage Hunter Ascendant do nothing, just follow my move, but don't know if is related
imagen

On outland it works really well
imagen

Also works find in other maps, love how they form a circle arround
imagen

**Comment by BarbzYHOOL** [comment]
the Mage in the Nexus not following also happens on old core, it's prob a problem with smart_scripts or the AI of the mob in cpp (depending where it's coded)

**Comment by locus313** [comment]
looking pretty good:
image

image

image

File: src/server/game/Entities/Unit/Unit.cpp [hunk 6]
```diff
@@ -153,9 +153,30 @@ ProcEventInfo::ProcEventInfo(Unit* actor, Unit* actionTarget, Unit* procTarget,
...
+    m_AutoRepeatFirstCast(false),
+    m_procDeep(0),
+    m_removedAurasCount(0),
+    i_motionMaster(new MotionMaster(this)),
+    m_regenTimer(0),
+    m_ThreatManager(this),
+    m_vehicle(nullptr),
+    m_vehicleKit(nullptr),
+    m_unitTypeMask(UNIT_MASK_NONE),
+    m_attackPosition(NULL),
```
Comment (locus313) [inline src/server/game/Entities/Unit/Unit.cpp:176]: The ubuntu-20.04-gcc build seems to pass if you change this NULL to false.

Comment (locus313) [reply src/server/game/Entities/Unit/Unit.cpp:176]: looks like it also resolved the ubuntu-18.04-gcc build error.

Comment (nicoaravena) [reply src/server/game/Entities/Unit/Unit.cpp:176]: Just compiled without errors and whiout any changes on Ubuntu 20.04

Comment (locus313) [reply src/server/game/Entities/Unit/Unit.cpp:176]: Right, it builds fine if the compiler is clang, but it is failing with gcc.

Comment (FrancescoBorzi) [reply src/server/game/Entities/Unit/Unit.cpp:176]: never use `NULL` as it's deprecated, use a modern counterpart:

- `nullptr` for pointers
- `0` for numbers
- `false` for booleans

File: src/server/game/Entities/Unit/Unit.cpp [hunk 7]
```diff
@@ -2189,6 +2225,128 @@ void Unit::AttackerStateUpdate (Unit* victim, WeaponAttackType attType, bool ext
...
+    AttackDistance(uint8 index, AttackPosition attackPos) : _index(index), _attackPos(attackPos) {}
+    uint8 _index;
+    AttackPosition _attackPos;
+};
+
+Position* Unit::GetMeleeAttackPoint(Unit* attacker)
+{
+    if (!attacker) // only player & pets to save CPU
+    {
+        return NULL;
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Unit/Unit.cpp:2238]: same

File: src/server/game/Entities/Unit/Unit.cpp [hunk 8]
```diff
@@ -2189,6 +2225,128 @@ void Unit::AttackerStateUpdate (Unit* victim, WeaponAttackType attType, bool ext
...
+    }
+
+    AttackerSet attackers = getAttackers();
+
+    if (attackers.size() <= 1) // if the attackers are not more than one
+    {
+        return NULL;
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Unit/Unit.cpp:2245]: same

File: src/server/game/Entities/Unit/Unit.cpp [hunk 9]
```diff
@@ -2189,6 +2225,128 @@ void Unit::AttackerStateUpdate (Unit* victim, WeaponAttackType attType, bool ext
...
+    }
+
+    float meleeReach = attacker->GetMeleeReach();
+
+    if (meleeReach <= 0)
+    {
+        return NULL;
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Unit/Unit.cpp:2252]: same

File: src/server/game/Entities/Unit/Unit.cpp [hunk 10]
```diff
@@ -2189,6 +2225,128 @@ void Unit::AttackerStateUpdate (Unit* victim, WeaponAttackType attType, bool ext
...
+
+    float attackerSize = attacker->GetObjectSize();
+
+    // in instance: the more attacker there are, the higher will be the tollerance
+    // outside: creatures should not intersecate
+    float distanceTollerance = attacker->GetMap()->IsDungeon() ? -attackerSize * tanh(validAttackers / 5.0f) : 0.0f;
+
+    if (!refUnit || minDistance > distanceTollerance)
+    {
+        return NULL;
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Unit/Unit.cpp:2291]: same

File: src/server/game/Entities/Unit/Unit.cpp [hunk 11]
```diff
@@ -2189,6 +2225,128 @@ void Unit::AttackerStateUpdate (Unit* victim, WeaponAttackType attType, bool ext
...
+        double victimY = GetPositionY();
+
+        // calculate tangent star
+        double a = 4.0f * ( pow(ray,2.0f) - pow(refUnitX,2.0f) + (2.0f * refUnitX * victimX) - pow(victimX,2.0f) );
+        double b = 8.0f * ( (refUnitX * refUnitY) + (victimX * victimY) - (victimX * refUnitY) - (refUnitX * victimY) );
+        double c = 4.0f * (- pow(victimY,2.0f) - pow(refUnitY,2.0f) + (2.0f*victimY*refUnitY) + pow(ray,2.0f));
+
+        if (a == 0) // should not happen
+        {
+            return NULL;
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Unit/Unit.cpp:2314]: same

File: src/server/game/Entities/Unit/Unit.h [hunk 12]
```diff
@@ -1255,6 +1255,26 @@ struct CharmInfo
...
+    };
+    int operator=(const int val)
+    {
+        if (!val)
+        {
+            // _pos = NULL;
+            _taken = false;
+            return 0; // NULL
+        }
+        return 0; // NULL
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Unit/Unit.h:1272]: same x3

File: src/server/game/Entities/Unit/Unit.h [hunk 13]
```diff
@@ -1405,6 +1425,7 @@ class Unit : public WorldObject
     virtual void SetCanDualWield(bool value) { m_canDualWield = value; }
     [[nodiscard]] float GetCombatReach() const override { return m_floatValues[UNIT_FIELD_COMBATREACH]; }
     [[nodiscard]] float GetMeleeReach() const { float reach = m_floatValues[UNIT_FIELD_COMBATREACH]; return reach > MIN_MELEE_REACH ? reach : MIN_MELEE_REACH; }
+    bool IsWithinRange(Unit const* obj, float dist) const;
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Unit/Unit.h:1428]: use `[[nodiscard]]` where appropriate

File: src/server/game/Entities/Unit/Unit.h [hunk 14]
```diff
@@ -1434,6 +1455,7 @@ class Unit : public WorldObject
     bool AttackStop();
     void RemoveAllAttackers();
     [[nodiscard]] AttackerSet const& getAttackers() const { return m_attackers; }
+    Position* GetMeleeAttackPoint(Unit* attacker);
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Unit/Unit.h:1458]: use `[[nodiscard]]` where appropriate

File: src/server/game/Entities/Unit/Unit.h [hunk 15]
```diff
@@ -2174,7 +2196,7 @@ class Unit : public WorldObject
 
     uint32 GetDisplayId() { return GetUInt32Value(UNIT_FIELD_DISPLAYID); }
     virtual void SetDisplayId(uint32 modelId);
-    uint32 GetNativeDisplayId() { return GetUInt32Value(UNIT_FIELD_NATIVEDISPLAYID); }
+    uint32 GetNativeDisplayId() const { return GetUInt32Value(UNIT_FIELD_NATIVEDISPLAYID); }
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Unit/Unit.h:2199]: use `[[nodiscard]]` where appropriate

File: src/server/game/Movement/MotionMaster.cpp [hunk 16]
```diff
@@ -320,12 +320,77 @@ void MotionMaster::MoveChase(Unit* target, float dist, float angle)
...
+}
+
+void MotionMaster::MoveCircleTarget(Unit* target)
+{
+    if (!target) {
+        return;
+    }
+
+    Position* point = target->GetMeleeAttackPoint(_owner);
+    if (point == NULL) {
```
Comment (FrancescoBorzi) [inline src/server/game/Movement/MotionMaster.cpp:357]: same

File: src/server/game/Movement/Spline/MoveSplineInit.cpp [hunk 17]
```diff
@@ -79,7 +79,10 @@ namespace Movement
         move_spline.onTransport = transport;
 
         uint32 moveFlags = unit->m_movementInfo.GetMovementFlags();
-        moveFlags |= (MOVEMENTFLAG_SPLINE_ENABLED | MOVEMENTFLAG_FORWARD);
+        //if (!args.flags.OrientationInversed)
+            moveFlags |= (MOVEMENTFLAG_SPLINE_ENABLED | MOVEMENTFLAG_FORWARD);
+        //else
+        //    moveFlags |= (MOVEMENTFLAG_SPLINE_ENABLED | MOVEMENTFLAG_BACKWARD);
```
Comment (FrancescoBorzi) [inline src/server/game/Movement/Spline/MoveSplineInit.cpp:85]: what about this?

File: src/common/Utilities/Geometry.h [hunk 18]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (C) 2016+     AzerothCore <www.azerothcore.org>, released under GNU GPL v2 license: https://github.com/azerothcore/azerothcore-wotlk/blob/master/LICENSE-GPL2
+ * Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/>
+ * Copyright (C) 2005-2009 MaNGOS <http://getmangos.com/>
+ */
```
Comment (FrancescoBorzi) [inline src/common/Utilities/Geometry.h:5]:
```cpp
/*
 * Copyright (C) 2016+     AzerothCore <www.azerothcore.org>, released under GNU AGPL v3 license: https://github.com/azerothcore/azerothcore-wotlk/blob/master/LICENSE-AGPL3
 */
```

File: src/common/Utilities/Geometry.h [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+#include <cmath>
+#include <iostream>
+#include <cstdlib>
+
+using namespace std;
+
+inline float getAngle(float startX, float startY, float destX, float destY)
+{
+    float dx = destX - startX;
+    float dy = destY - startY;
```
Comment (FrancescoBorzi) [inline src/common/Utilities/Geometry.h:19]: you can use `auto` instead of `float`

File: src/common/Utilities/Geometry.h [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+
+    float ang = atan2(dy, dx);
```
Comment (FrancescoBorzi) [inline src/common/Utilities/Geometry.h:19]: same

File: src/common/Utilities/Geometry.h [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+    return ang;
+}
+
+inline float getSlopeAngle(float startX, float startY, float startZ, float destX, float destY, float destZ)
+{
+    float a = (startX * destX + startY * destY + startZ * destZ);
+    float b = sqrt(pow(startX,2.0f) + pow(startY,2.0f) + pow(startZ, 2.0f));
+    float c = sqrt(pow(destX,2.0f) + pow(destY,2.0f) + pow(destZ, 2.0f));
+
+    float ang = acos(a / (b * c));
```
Comment (FrancescoBorzi) [inline src/common/Utilities/Geometry.h:30]: same

File: src/common/Utilities/Geometry.h [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+
+    if (isnan(ang)) {
```
Comment (FrancescoBorzi) [inline src/common/Utilities/Geometry.h:32]: codestyle (this is not TypeScript :D)

**Review by FrancescoBorzi — APPROVED** [review]

**Review by Yehonal — APPROVED** [review]
I'm going to create another pull request to improve the Pathfinding adding runtime check of the slopes and several other optimizations. But this specific PR must be merged asap

**Comment by pklloveyou** [comment]
Why is it so complicated，Just delete this line of code
TargetedMovementGenerator.cpp
```c++
i_target->GetContactPoint(owner, x, y, z);
```

**Comment by Yehonal** [comment]
> ```c++
> i_target->GetContactPoint(
> ```

Good catch but...

Removing that line allow us to have a nicer positioning from the beginning, however, looks weird but that is not the retail behaviour.

On retail servers the creatures reach you in a cone/pile and then they start to move on left or right gradually to find their spot. 

Please check this video (these are my tests from retail) and tell me what you think: https://www.youtube.com/watch?v=ogmPCycMC3U&list=PLx1yhYXFOxyFCrLWJIr2dfyySx5jiEr4k&index=5

**Pull request merged**